### PR TITLE
Update Package : METIS/ParMETIS

### PR DESCRIPTION
**Pull request by xjrc**
The changes in this pull request make the following changes to the `metis` and `parmetis` packages:

- Add installation support for package versions `metis@5.0.2` and `parmetis@4.0.2`.
- Add a `url_for_version` function to each package to remove the need for per-version urls.
- Change the `metis+double` variant to `metis+real64` to improve naming consistency with the `metis+idx64` variant.
- Remove an unnecessary dependency in `metis` on the `gdb` package.
- Improve the `metis@4` installation method by adding support for working with staged archives, fixing a few Linux installation errors, and adding better install error handling.
- Update the code in these packages to conform to PEP8 standards.

I've verified that the `metis@4.0.3/5.0.2+/~shared` and `parmetis@4.0.2+/~shared` variants of these packages compile and install properly on CHAOS when compiling with `gcc@4.7.2`.

@goxberry, @davydden : Could you both verify that your installation variants for `metis` and `parmetis` still work before this PR is merged?  Thanks in advance for your help and please let me know if you have feedback on these changes.

### CODE REVIEW

File: var/spack/repos/builtin/packages/parmetis/package.py [hunk 1]
```diff
@@ -26,64 +26,64 @@
...
-        # FIXME : Once a contract is defined, MPI compilers should be retrieved indirectly via spec['mpi'] in case
-        # FIXME : they use a non-standard name
-        options.extend(['-DGKLIB_PATH:PATH={metis_source}/GKlib'.format(metis_source=spec['metis'].prefix.include),
-                        '-DMETIS_PATH:PATH={metis_source}'.format(metis_source=spec['metis'].prefix),
-                        '-DCMAKE_C_COMPILER:STRING=mpicc',
-                        '-DCMAKE_CXX_COMPILER:STRING=mpicxx'])
+        options.extend([
+            '-DGKLIB_PATH:PATH=%s/GKlib' % spec['metis'].prefix.include,
+            '-DMETIS_PATH:PATH=%s' % spec['metis'].prefix,
+            '-DCMAKE_C_COMPILER:STRING=%s' % spec['mpi'].mpicc,
```
Comment (davydden) [inline var/spack/repos/builtin/packages/parmetis/package.py]: nice!

**Comment by davydden** [comment]
i will test on osx now...

**Comment by davydden** [comment]
compiled fine on osx with Clang 7.3.0.

**Comment by xjrc** [comment]
@davydden: Excellent, thank you!

**Pull request merged**